android/library/src/main/java/com/liferay/mobile/screens/ddl/model/DocumentField.java

```diff
@@ -20,142 +20,155 @@
-    }
-
-    private enum State {
-        IDLE, UPLOADING, UPLOADED, FAILED
-    }
+	public static final Parcelable.ClassLoaderCreator<DocumentField> CREATOR =
+		new Parcelable.ClassLoaderCreator<DocumentField>() {
+
+			@Override
+			public DocumentField createFromParcel(Parcel parcel, ClassLoader classLoader) {
+				return new DocumentField(parcel, classLoader);
+			}
+
+			public DocumentField createFromParcel(Parcel in) {
+				throw new AssertionError();
+			}
+
+			public DocumentField[] newArray(int size) {
+				return new DocumentField[size];
+			}
+		};
+	private State state = State.IDLE;
+
+	public DocumentField() {
+		super();
+	}
+
+	public DocumentField(Map<String, Object> attributes, Locale locale, Locale defaultLocale) {
+		super(attributes, locale, defaultLocale);
+	}
+
+	protected DocumentField(Parcel in, ClassLoader classLoader) {
+		super(in, classLoader);
+	}
+
+	public boolean moveToUploadInProgressState() {
+		if (state == State.IDLE || state == State.FAILED || state == State.UPLOADED) {
+			state = State.UPLOADING;
+			return true;
+		}
+
+		return false;
+	}
+
+	public boolean moveToUploadCompleteState() {
+		if (state == State.UPLOADING) {
+			state = State.UPLOADED;
+			return true;
+		}
+
+		return false;
+	}
+
+	public boolean moveToUploadFailureState() {
+		if (state == State.UPLOADING) {
+			state = State.FAILED;
+			return true;
+		}
+
+		return false;
+	}
+
+	public boolean isUploaded() {
+		return (state == State.UPLOADED);
+	}
+
+	public boolean isUploading() {
+		return (state == State.UPLOADING);
+	}
+
+	public boolean isUploadFailed() {
+		return (state == State.FAILED);
+	}
+
+	public void createLocalFile(String path) {
+		if (path == null || path.isEmpty()) {
+			return;
+		}
+
+		setCurrentValue(new DocumentLocalFile(path));
+	}
+
+	@Override
+	protected DocumentFile convertFromString(String string) {
+		if (string == null || string.isEmpty() || isEmptyJson(string)) {
+			return null;
+		}
+
+		DocumentRemoteFile result = null;
+
+		try {
+			result = new DocumentRemoteFile(string);
+		} catch (JSONException e) {
+			LiferayLogger.e("Can't parse the document JSON", e);
+		}
+
+		return result;
+	}
+
+	@Override
+	protected String convertToData(DocumentFile document) {
+		return document == null ? null : document.toData();
+	}
+
+	@Override
+	protected String convertToFormattedString(DocumentFile document) {
+		return document == null ? "" : document.getFileName();
+	}
+
+	@Override
+	protected boolean doValidate() {
+		boolean valid = true;
+
+		DocumentFile currentValue = getCurrentValue();
+
+		if (currentValue == null) {
+			if (isRequired()) {
+				valid = false;
+			}
+		} else {
+			if (!currentValue.isValid()) {
+				valid = false;
+			}
+		}
+
+		if (isUploading()) {
+			valid = false;
+		} else if (isUploadFailed()) {
+			valid = false;
+		}
+
+		return valid;
+	}
+
+	private boolean isEmptyJson(String stringValue) {
+		try {
+			if (new JSONObject(stringValue).keys().hasNext()) {
```

Comment:
A json with keys is an empty Json?

Reply:
I think I'm missing some context here

Reply:
Victor yeah that logic was wrong. But I rebuild the solution using another approach.

Reply:
@victorg1991 I'll create a test for that